### ISSUE TITLE
Bug: Importing certain calenders results in "requires Javascript" error

### ISSUE THREAD

**Issue by jessechahal** [nextcloud/integration_google#29]
As with https://github.com/nextcloud/integration_google/issues/11 it seems like this bug is still affecting the latest version of the app.
I'm also getting tons of errors with photos import & calendar import. 
Some calendars imported fine (secondary calendars) but other calendars (main personal calendars) did not.

In the WebUI I'm still getting this bug _This application requires JavaScript for correct operation. Please enable JavaScript and reload the page_ when trying to import calendars. In the Nextcloud logs I'm seeing all types of errors.

```
Nextcloud version: 20.0.1.1
google_plugin_version: 0.15
Ubuntu 20.04LTS on LXC
OS HDD size: 8GB
Nextcloud_data_hdd: 50GB
```

**Nextcloud related log message:**
Calendar error(s)

PostgresSQL error
```
{"reqId":"2mWfZKDbkHjjwpJWqtlJ","level":3,"time":"2020-11-07T02:15:49+00:00","remoteAddr":"1.2.3.4","user":"jesse","app":"index","method":"GET","url":"/index.php/apps/integration_google/import-calendar?calId=email%40gmail.com&calName=email%40gmail.com&color=%239fc6e7","message":{"Exception":"Doctrine\\DBAL\\Exception\\DriverException","Message":"An exception occurred while executing 'INSERT INTO \"oc_calendarobjects_props\" (\"calendarid\", \"calendartype\", \"objectid\", \"name\", \"parameter\", \"value\") VALUES(?, ?, ?, ?, ?, ?)' with params [18, 0, 2739, \"DESCRIPTION\", null, \"Hello Jesse,\\\\n\\\\n\\\\nThanks for your continue interest in CompanyX and the Application Developer role.\\\\n\\\\n\\\\nYou will be meeting with the following individuals tomorrow at 4pm.\\\\n\\\\n\\\\nLead, Application Developer \\u2013 Clark Kent\\\\nTechnical Lead \\u2013 G\"]:\n\nSQLSTATE[22001]: String data, right truncated: 7 ERROR:  value too long for type character varying(255)","Code":0,"Trace":[{"file":"/var/www/nextcloud/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/DBALException.php","line":169,"function":"convertException","class":"Doctrine\\DBAL\\Driver\\AbstractPostgreSQLDriver","type":"->"},{"file":"/var/www/nextcloud/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/DBALException.php","line":149,"function":"wrapException","class":"Doctrine\\DBAL\\DBALException","type":"::"},{"file":"/var/www/nextcloud/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Connection.php","line":1071,"function":"driverExceptionDuringQuery","class":"Doctrine\\DBAL\\DBALException","type":"::"},{"file":"/var/www/nextcloud/lib/private/DB/Connection.php","line":214,"function":"executeUpdate","class":"Doctrine\\DBAL\\Connection","type":"->"},{"file":"/var/www/nextcloud/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Query/QueryBuilder.php","line":209,"function":"executeUpdate","class":"OC\\DB\\Connection","type":"->"},{"file":"/var/www/nextcloud/lib/private/DB/QueryBuilder/QueryBuilder.php","line":217,"function":"execute","class":"Doctrine\\DBAL\\Query\\QueryBuilder","type":"->"},{"file":"/var/www/nextcloud/apps/dav/lib/CalDAV/CalDavBackend.php","line":2619,"function":"execute","class":"OC\\DB\\QueryBuilder\\QueryBuilder","type":"->"},{"file":"/var/www/nextcloud/apps/dav/lib/CalDAV/CalDavBackend.php","line":1117,"function":"updateProperties","class":"OCA\\DAV\\CalDAV\\CalDavBackend","type":"->","args":["*** sensitive parameters replaced ***"]},{"file":"/var/www/nextcloud/apps/integration_google/lib/Service/GoogleCalendarAPIService.php","line":167,"function":"createCalendarObject","class":"OCA\\DAV\\CalDAV\\CalDavBackend","type":"->"},{"file":"/var/www/nextcloud/apps/integration_google/lib/Controller/GoogleAPIController.php","line":229,"function":"importCalendar","class":"OCA\\Google\\Service\\GoogleCalendarAPIService","type":"->"},{"file":"/var/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php","line":169,"function":"importCalendar","class":"OCA\\Google\\Controller\\GoogleAPIController","type":"->"},{"file":"/var/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php","line":100,"function":"executeController","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->"},{"file":"/var/www/nextcloud/lib/private/AppFramework/App.php","line":152,"function":"dispatch","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->"},{"file":"/var/www/nextcloud/lib/private/Route/Router.php","line":308,"function":"main","class":"OC\\AppFramework\\App","type":"::"},{"file":"/var/www/nextcloud/lib/base.php","line":1009,"function":"match","class":"OC\\Route\\Router","type":"->"},{"file":"/var/www/nextcloud/index.php","line":37,"function":"handleRequest","class":"OC","type":"::"}],"File":"/var/www/nextcloud/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractPostgreSQLDriver.php","Line":79,"Previous":{"Exception":"Doctrine\\DBAL\\Driver\\PDOException","Message":"SQLSTATE[22001]: String data, right truncated: 7 ERROR:  value too long for type character varying(255)","Code":"22001","Trace":[{"file":"/var/www/nextcloud/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Connection.php","line":1062,"function":"execute","class":"Doctrine\\DBAL\\Driver\\PDOStatement","type":"->"},{"file":"/var/www/nextcloud/lib/private/DB/Connection.php","line":214,"function":"executeUpdate","class":"Doctrine\\DBAL\\Connection","type":"->"},{"file":"/var/www/nextcloud/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Query/QueryBuilder.php","line":209,"function":"executeUpdate","class":"OC\\DB\\Connection","type":"->"},{"file":"/var/www/nextcloud/lib/private/DB/QueryBuilder/QueryBuilder.php","line":217,"function":"execute","class":"Doctrine\\DBAL\\Query\\QueryBuilder","type":"->"},{"file":"/var/www/nextcloud/apps/dav/lib/CalDAV/CalDavBackend.php","line":2619,"function":"execute","class":"OC\\DB\\QueryBuilder\\QueryBuilder","type":"->"},{"file":"/var/www/nextcloud/apps/dav/lib/CalDAV/CalDavBackend.php","line":1117,"function":"updateProperties","class":"OCA\\DAV\\CalDAV\\CalDavBackend","type":"->","args":["*** sensitive parameters replaced ***"]},{"file":"/var/www/nextcloud/apps/integration_google/lib/Service/GoogleCalendarAPIService.php","line":167,"function":"createCalendarObject","class":"OCA\\DAV\\CalDAV\\CalDavBackend","type":"->"},{"file":"/var/www/nextcloud/apps/integration_google/lib/Controller/GoogleAPIController.php","line":229,"function":"importCalendar","class":"OCA\\Google\\Service\\GoogleCalendarAPIService","type":"->"},{"file":"/var/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php","line":169,"function":"importCalendar","class":"OCA\\Google\\Controller\\GoogleAPIController","type":"->"},{"file":"/var/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php","line":100,"function":"executeController","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->"},{"file":"/var/www/nextcloud/lib/private/AppFramework/App.php","line":152,"function":"dispatch","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->"},{"file":"/var/www/nextcloud/lib/private/Route/Router.php","line":308,"function":"main","class":"OC\\AppFramework\\App","type":"::"},{"file":"/var/www/nextcloud/lib/base.php","line":1009,"function":"match","class":"OC\\Route\\Router","type":"->"},{"file":"/var/www/nextcloud/index.php","line":37,"function":"handleRequest","class":"OC","type":"::"}],"File":"/var/www/nextcloud/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php","Line":129,"Previous":{"Exception":"PDOException","Message":"SQLSTATE[22001]: String data, right truncated: 7 ERROR:  value too long for type character varying(255)","Code":"22001","Trace":[{"file":"/var/www/nextcloud/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php","line":127,"function":"execute","class":"PDOStatement","type":"->"},{"file":"/var/www/nextcloud/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Connection.php","line":1062,"function":"execute","class":"Doctrine\\DBAL\\Driver\\PDOStatement","type":"->"},{"file":"/var/www/nextcloud/lib/private/DB/Connection.php","line":214,"function":"executeUpdate","class":"Doctrine\\DBAL\\Connection","type":"->"},{"file":"/var/www/nextcloud/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Query/QueryBuilder.php","line":209,"function":"executeUpdate","class":"OC\\DB\\Connection","type":"->"},{"file":"/var/www/nextcloud/lib/private/DB/QueryBuilder/QueryBuilder.php","line":217,"function":"execute","class":"Doctrine\\DBAL\\Query\\QueryBuilder","type":"->"},{"file":"/var/www/nextcloud/apps/dav/lib/CalDAV/CalDavBackend.php","line":2619,"function":"execute","class":"OC\\DB\\QueryBuilder\\QueryBuilder","type":"->"},{"file":"/var/www/nextcloud/apps/dav/lib/CalDAV/CalDavBackend.php","line":1117,"function":"updateProperties","class":"OCA\\DAV\\CalDAV\\CalDavBackend","type":"->","args":["*** sensitive parameters replaced ***"]},{"file":"/var/www/nextcloud/apps/integration_google/lib/Service/GoogleCalendarAPIService.php","line":167,"function":"createCalendarObject","class":"OCA\\DAV\\CalDAV\\CalDavBackend","type":"->"},{"file":"/var/www/nextcloud/apps/integration_google/lib/Controller/GoogleAPIController.php","line":229,"function":"importCalendar","class":"OCA\\Google\\Service\\GoogleCalendarAPIService","type":"->"},{"file":"/var/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php","line":169,"function":"importCalendar","class":"OCA\\Google\\Controller\\GoogleAPIController","type":"->"},{"file":"/var/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php","line":100,"function":"executeController","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->"},{"file":"/var/www/nextcloud/lib/private/AppFramework/App.php","line":152,"function":"dispatch","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->"},{"file":"/var/www/nextcloud/lib/private/Route/Router.php","line":308,"function":"main","class":"OC\\AppFramework\\App","type":"::"},{"file":"/var/www/nextcloud/lib/base.php","line":1009,"function":"match","class":"OC\\Route\\Router","type":"->"},{"file":"/var/www/nextcloud/index.php","line":37,"function":"handleRequest","class":"OC","type":"::"}],"File":"/var/www/nextcloud/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php","Line":127}},"CustomMessage":"--"},"userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.111 Safari/537.36","version":"20.0.1.1","id":"5fa61b8ea4a80"}
```

this is from a calendar shared to my google account from gsuite account
```
{"reqId":"NI8PpKLaiuXLZhrqG3v6","level":3,"time":"2020-11-07T02:11:53+00:00","remoteAddr":"1.2.3.4","user":"jesse","app":"PHP","method":"GET","url":"/index.php/apps/integration_google/import-calendar?calId=email%example.com&calName=email%example.com&color=%2342d692","message":{"Exception":"Error","Message":"Undefined variable: created at /var/www/nextcloud/apps/integration_google/lib/Service/GoogleCalendarAPIService.php#107","Code":0,"Trace":[{"file":"/var/www/nextcloud/apps/integration_google/lib/Service/GoogleCalendarAPIService.php","line":107,"function":"onError","class":"OC\\Log\\ErrorHandler","type":"::"},{"file":"/var/www/nextcloud/apps/integration_google/lib/Controller/GoogleAPIController.php","line":229,"function":"importCalendar","class":"OCA\\Google\\Service\\GoogleCalendarAPIService","type":"->"},{"file":"/var/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php","line":169,"function":"importCalendar","class":"OCA\\Google\\Controller\\GoogleAPIController","type":"->"},{"file":"/var/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php","line":100,"function":"executeController","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->"},{"file":"/var/www/nextcloud/lib/private/AppFramework/App.php","line":152,"function":"dispatch","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->"},{"file":"/var/www/nextcloud/lib/private/Route/Router.php","line":308,"function":"main","class":"OC\\AppFramework\\App","type":"::"},{"file":"/var/www/nextcloud/lib/base.php","line":1009,"function":"match","class":"OC\\Route\\Router","type":"->"},{"file":"/var/www/nextcloud/index.php","line":37,"function":"handleRequest","class":"OC","type":"::"}],"File":"/var/www/nextcloud/lib/private/Log/ErrorHandler.php","Line":91,"CustomMessage":"--"},"userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.111 Safari/537.36","version":"20.0.1.1","id":"5fa61b8ea4b90"}
```
Caldev error?
```
{"reqId":"iBxBHknoi2plRoK7l8vB","level":4,"time":"2020-11-07T02:04:12+00:00","remoteAddr":"1.2.3.4","user":"jesse","app":"webdav","method":"PUT","url":"/remote.php/dav/addressbooks/users/jesse/contacts/goog153","message":{"Exception":"Sabre\\DAV\\Exception\\UnsupportedMediaType","Message":"Validation error in vCard: ENCODING parameter is not valid in vCard 4.","Code":0,"Trace":[{"file":"/var/www/nextcloud/3rdparty/sabre/dav/lib/CardDAV/Plugin.php","line":282,"function":"validateVCard","class":"Sabre\\CardDAV\\Plugin","type":"->"},{"file":"/var/www/nextcloud/3rdparty/sabre/event/lib/WildcardEmitterTrait.php","line":89,"function":"beforeWriteContent","class":"Sabre\\CardDAV\\Plugin","type":"->"},{"file":"/var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php","line":1139,"function":"emit","class":"Sabre\\DAV\\Server","type":"->"},{"file":"/var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/CorePlugin.php","line":515,"function":"updateFile","class":"Sabre\\DAV\\Server","type":"->","args":["*** sensitive parameters replaced ***"]},{"file":"/var/www/nextcloud/3rdparty/sabre/event/lib/WildcardEmitterTrait.php","line":89,"function":"httpPut","class":"Sabre\\DAV\\CorePlugin","type":"->"},{"file":"/var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php","line":474,"function":"emit","class":"Sabre\\DAV\\Server","type":"->"},{"file":"/var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php","line":251,"function":"invokeMethod","class":"Sabre\\DAV\\Server","type":"->"},{"file":"/var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php","line":319,"function":"start","class":"Sabre\\DAV\\Server","type":"->"},{"file":"/var/www/nextcloud/apps/dav/lib/Server.php","line":325,"function":"exec","class":"Sabre\\DAV\\Server","type":"->"},{"file":"/var/www/nextcloud/apps/dav/appinfo/v2/remote.php","line":35,"function":"exec","class":"OCA\\DAV\\Server","type":"->"},{"file":"/var/www/nextcloud/remote.php","line":167,"args":["/var/www/nextcloud/apps/dav/appinfo/v2/remote.php"],"function":"require_once"}],"File":"/var/www/nextcloud/3rdparty/sabre/dav/lib/CardDAV/Plugin.php","Line":376,"CustomMessage":"--"},"userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.111 Safari/537.36","version":"20.0.1.1","id":"5fa61b8ea4c48"}
```

**Photos error(s):**
```
{"reqId":"ofCshcW8p1J1WqxYuWHr","level":3,"time":"2020-11-07T02:49:47+00:00","remoteAddr":"1.2.3.4","user":"jesse","app":"PHP","method":"GET","url":"/index.php/apps/integration_google/photo-number","message":{"Exception":"Error","Message":"Undefined index: mediaItems at /var/www/nextcloud/apps/integration_google/lib/Service/GooglePhotosAPIService.php#67","Code":0,"Trace":[{"file":"/var/www/nextcloud/apps/integration_google/lib/Service/GooglePhotosAPIService.php","line":67,"function":"onError","class":"OC\\Log\\ErrorHandler","type":"::"},{"file":"/var/www/nextcloud/apps/integration_google/lib/Controller/GoogleAPIController.php","line":118,"function":"getPhotoNumber","class":"OCA\\Google\\Service\\GooglePhotosAPIService","type":"->"},{"file":"/var/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php","line":169,"function":"getPhotoNumber","class":"OCA\\Google\\Controller\\GoogleAPIController","type":"->"},{"file":"/var/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php","line":100,"function":"executeController","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->"},{"file":"/var/www/nextcloud/lib/private/AppFramework/App.php","line":152,"function":"dispatch","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->"},{"file":"/var/www/nextcloud/lib/private/Route/Router.php","line":308,"function":"main","class":"OC\\AppFramework\\App","type":"::"},{"file":"/var/www/nextcloud/lib/base.php","line":1009,"function":"match","class":"OC\\Route\\Router","type":"->"},{"file":"/var/www/nextcloud/index.php","line":37,"function":"handleRequest","class":"OC","type":"::"}],"File":"/var/www/nextcloud/lib/private/Log/ErrorHandler.php","Line":91,"CustomMessage":"--"},"userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.111 Safari/537.36","version":"20.0.1.1","id":"5fa61b8ea4589"}
```
quota error
```
{"reqId":"DJCADohKswT3ktIF9nLA","level":2,"time":"2020-11-07T02:16:26+00:00","remoteAddr":"1.2.3.4","user":"jesse","app":"integration_google","method":"GET","url":"/index.php/apps/integration_google/photo-number","message":"Google API error : Client error: `GET https://photoslibrary.googleapis.com/v1/mediaItems?pageSize=100&pageToken=XXXXXXXXXblahblahHERExxxxxxxxxxx` resulted in a `429 Too Many Requests` response:\n{\n  \"error\": {\n    \"code\": 429,\n    \"message\": \"Quota exceeded for quota metric 'Read requests' and limit 'Read requests (truncated...)\n","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.111 Safari/537.36","version":"20.0.1.1","id":"5fa61b8ea48a9"}
```

Other things I'm worried about is the fact my photos are 10GB+ and drive/docs data is 10GB+ but my main hdd is only 8GB. From what i read somewhere `/tmp` folder is used to copy data over. If the migration waits until all data is copied from google before moving over to nextcloud_data (hdd is 50GB) then it'll never work.

**Comment by eneiluj** [comment]
Wow that's a great bug report. Thanks a lot.

> PostgresSQL error

Field values are now truncated to fill the 255 character db field. Should be fine.

> this is from a calendar shared to my google account from gsuite account

Stupid mistake, fixed.

 > Caldev error?

I don't know what's happening there. I made it safer, it should not crash anymore and just log when a contact image couldn't be imported.

> Photos error(s):

This is strange. I don't know why the hell google API returns an incorrect value. Anyway I made it safer, it should not crash anymore and just ignore those kind of responses.

> quota error

Well this can't really be avoided if you make too many page refresh. I guess you have many many photos and the way the app gets the number of photos is pretty bad (didn't find a nice way with the API). The app actually gets a list of all photo names by chunks of 50. So if you have N photos, it makes `N / 50` requests. That's why you face the nice `429 Too Many Requests` response.

So, v0.0.17 is out. It would be awesome if you could try it.

About your disk space concern, no worries, even when temp files were created, each one was deleted after having copied it to your actual storage. But this was before, files/photos are now directly downloaded to the storage (thanks to the 'sink' option of Guzzle http client). I'm also very interested by your feedback with a real life big drive/photo import. Specially if you have big files to check if Guzzle does not take too much RAM (and hit Php memory limit) when downloading them.

Thank you very much for your help in solving those (very abundant) bugs :rocket:

**Comment by eneiluj** [comment]
I found a way to reduce the number of requests when getting photo number. Problem is it does not get photos that don't belong to an album. So the photo number is approximate, anyway the total photo collection size was already approximate.

Now If you have N albums, you make `N / 50` requests to get the photo number. I hope it solves the request quota issue. It's in v0.0.18 which is out.

**Comment by jessechahal** [comment]
Thanks for such a quick update, I thought it might take weeks to get looked at! I've upgraded to v0.18 and the results are below.

### Importing calendars status:
- Personal gmail calendar works fine! 👍 
- GSuite calendar shared to me
  - all events get imported but looking at the events in NextCloud shows that none of them have titles. They show up as "Untitled event". 
    - this isn't really a high priority calendar for me, just thought i would mention whats happening. Not sure if its important but all events are ones that have occurred in the past thus why this calendar isn't too important to me. 
    - no error logs in NextCloud logs
- Side non-issue with calendar imports.
  - If you have a dead CalDev calendar in Google Calendar it fails to import (as it should fail) but the error msg in the UI is just a straight dump of client side error. Maybe throw up a more user friendly error and direct enduser to look at nextcloud logs for more information? I have a facebook birthday calendar shared with my google account. At some point facebook killed support for this (i set it up in 2010). Old events still show up in google calendar (probably due to some caching of events that occur yearly) but ofcourse the actual facebook CalDev calendar is dead. 

```
{"reqId":"v2ozwPZhMZBGzESdz6ZU","level":2,"time":"2020-11-08T22:56:48+00:00","remoteAddr":"1.2.3.4","user":"jesse","app":"integration_google","method":"GET","url":"/index.php/apps/integration_google/import-calendar?calId=addressbook%23contacts%40group.v.calendar.google.com&calName=Birthdays&color=%239a9cff","message":"Google API error : Client error: `GET https://www.googleapis.com/calendar/v3/calendars/addressbook#contacts@group.v.calendar.google.com/events?maxResults=100` resulted in a `404 Not Found` response:\n{\n \"error\": {\n  \"errors\": [\n   {\n    \"domain\": \"global\",\n    \"reason\": \"notFound\",\n    \"message\": \"Not Found\"\n   }\n  ],\n (truncated...)\n","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.111 Safari/537.36","version":"20.0.1.1","id":"5fa877b1217e6"}
```

### Google Photos import
- Correctly enumerates the number of files and total size
- seems stuck during the actual import process. See google drive log message below. I believe it might somehow be affects photos import as well since the same error message appears in logs whether i try to import photos or drive files. 

### Google Drive import
- GUI shows "1000 files in Google Drive (13 GiB)"
  - 1000 sees like an oddly round number, i probably have either more or less than 1000. 
- seems frozen on data import page, never ends up importing anything
```
{"reqId":"oTOFARVPOIRCuRdeWLXX","level":3,"time":"2020-11-08T23:16:23+00:00","remoteAddr":"1.2.3.4","user":"jesse","app":"PHP","method":"GET","url":"/index.php/apps/integration_google/drive-size","message":{"Exception":"Error","Message":"Undefined index: size at /var/www/nextcloud/apps/integration_google/lib/Service/GoogleDriveAPIService.php#85","Code":0,"Trace":[{"file":"/var/www/nextcloud/apps/integration_google/lib/Service/GoogleDriveAPIService.php","line":85,"function":"onError","class":"OC\\Log\\ErrorHandler","type":"::"},{"file":"/var/www/nextcloud/apps/integration_google/lib/Controller/GoogleAPIController.php","line":172,"function":"getDriveSize","class":"OCA\\Google\\Service\\GoogleDriveAPIService","type":"->"},{"file":"/var/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php","line":169,"function":"getDriveSize","class":"OCA\\Google\\Controller\\GoogleAPIController","type":"->"},{"file":"/var/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php","line":100,"function":"executeController","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->"},{"file":"/var/www/nextcloud/lib/private/AppFramework/App.php","line":152,"function":"dispatch","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->"},{"file":"/var/www/nextcloud/lib/private/Route/Router.php","line":308,"function":"main","class":"OC\\AppFramework\\App","type":"::"},{"file":"/var/www/nextcloud/lib/base.php","line":1009,"function":"match","class":"OC\\Route\\Router","type":"->"},{"file":"/var/www/nextcloud/index.php","line":37,"function":"handleRequest","class":"OC","type":"::"}],"File":"/var/www/nextcloud/lib/private/Log/ErrorHandler.php","Line":91,"CustomMessage":"--"},"userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.111 Safari/537.36","version":"20.0.1.1","id":"5fa87c5708b52"}
```

#### Other possibly useful info:
PHP settings
```
Version: 7.4.3
Memory Limit: 512 MB
Max Execution Time: 3600
Upload max size: 2 MB
```
- I'm a bit worried about that max upload size during migration, its easily possible to have files in google drive that are larger then 2MB. If it becomes an issue I can probably change the PHP settings somewhere for this but it'll probably be a problem for more average users (as far as i'm aware NextCloud best practice settings make no mention of 'Upload max size'). 
- I'm still far from hitting the configured PHP memory limit but of course since importing of photos and drive/docs files seems to be blocking on something else this might still somehow become an issue during the actual import.

**Comment by eneiluj** [comment]
* "Untitled event" issue: I don't see any possible bug there. The only reason for that is that there is nothing in the "summary" field for those events in the JSON data received. As I can't register to GSuite for free (they say it's for free but a credit card number is mandatory...) I can't try to share such calendar with my test google user. If you can share a GSuite calendar with me (nyugglee@gmail.com) I could inspect the difference in the API response when importing.

* > Side non-issue with calendar imports

Interesting. I would rather avoid listing those calendars as they are not really google ones. I'll see if it's possible to make the difference between classic and subscribed calendars when getting them with the API.

* Photo and Drive import issue: Solved in v0.0.19. This API is driving me crazy :grin:. I don't know why the size is not there for some files. It's done in a safer way now.

> I'm a bit worried about that max upload size

Don't worry, this limit does not apply here as you're not uploading anything but asking Nextcloud server to download stuff.

* PHP memory limit: Don't worry there too. Files are not stored in memory but progressively written to target files.

**Comment by jessechahal** [comment]
I've updated to `v0.19`. Photos importing looking good so far?

### Background jobs
I think a large part of the reason the photos migration was not working on my system (staying at 0 files imported) was due to misconfigured background jobs. Originally it was setup with the default out-of-the-box AJAX config (worst possible option). 

I reconfigured _<nextcloud_url>/index.php/settings/admin_ to use cron following the documentation at [https://docs.nextcloud.com/server/20/admin_manual/configuration_server/background_jobs_configuration.html](https://docs.nextcloud.com/server/20/admin_manual/configuration_server/background_jobs_configuration.html). I am now at 144% imported for photos and still increasing.

During photo import I'm seeing the below error in the logs. My guess is the stream is closed by something else by the time fclose() is called on the file/photo which causes close-on-invalid-stream error in the log. If trying to call fclose() on something gotten through google's API using RESTful calls (aka HTTP) i assume the stream would automatically be closed once the HTTP request has completed (unless using one of those newer fangled HTTP2/HTTP3 libraries to make calls). 
```
{"reqId":"uCTnOCxwVURCKrj003Hs","level":3,"time":"2020-11-09T02:24:45+00:00","remoteAddr":"","user":"--","app":"PHP","method":"","url":"--","message":{"Exception":"Error","Message":"fclose(): supplied resource is not a valid stream resource at /var/www/nextcloud/apps/integration_google/lib/Service/GooglePhotosAPIService.php#334","Code":0,"Trace":[{"function":"onError","class":"OC\\Log\\ErrorHandler","type":"::"},{"file":"/var/www/nextcloud/apps/integration_google/lib/Service/GooglePhotosAPIService.php","line":334,"function":"fclose"},{"file":"/var/www/nextcloud/apps/integration_google/lib/Service/GooglePhotosAPIService.php","line":253,"function":"getPhoto","class":"OCA\\Google\\Service\\GooglePhotosAPIService","type":"->"},{"file":"/var/www/nextcloud/apps/integration_google/lib/Service/GooglePhotosAPIService.php","line":143,"function":"importPhotos","class":"OCA\\Google\\Service\\GooglePhotosAPIService","type":"->"},{"file":"/var/www/nextcloud/apps/integration_google/lib/BackgroundJob/ImportPhotosJob.php","line":36,"function":"importPhotosJob","class":"OCA\\Google\\Service\\GooglePhotosAPIService","type":"->"},{"file":"/var/www/nextcloud/lib/public/BackgroundJob/Job.php","line":80,"function":"run","class":"OCA\\Google\\BackgroundJob\\ImportPhotosJob","type":"->"},{"file":"/var/www/nextcloud/lib/public/BackgroundJob/QueuedJob.php","line":48,"function":"execute","class":"OCP\\BackgroundJob\\Job","type":"->"},{"file":"/var/www/nextcloud/cron.php","line":127,"function":"execute","class":"OCP\\BackgroundJob\\QueuedJob","type":"->"}],"File":"/var/www/nextcloud/lib/private/Log/ErrorHandler.php","Line":91,"CustomMessage":"--"},"userAgent":"--","version":"20.0.1.1","id":"5fa8a86d52d73"}
```

### Photos import

So there is a bit of weirdness here. Currently it is showing 144% imported and still increasing. This calculation is probably based on the fact that I also imported shared albums which made up ~25-40% of the uploaded pictures. The odd things are that the with `v0.14` photo importer guestimated that i had 13GB of photos (could have been mistake since it also said drive/docs was 13GB too). With `v0.15`+ it estimated 4GB. According to nextcloud the photos on disk currently take up 484MB of disk space. Honestly i expect the photos to be in gigabytes because I used to have original PixelXL phone (came with free unlimited google photo storage at original quality). The process is still going on in the background so possible photos on disk will eventually be in GB's but the completely process will probably end up at 1000%+ 

**Question**: how is photo download quality decided?

### Drive import

Folder structure has imported fine. Some of the files were imported and converted. When i try to open the files they fail to open complaining about issues with URL (local ipv4 instead of DNS record). Creating new files in nextcloud causes them to correctly use DNS record. I think there might still be a few issues on my side on this so going to wait for photo import to completely finish before trying to evalute drive/docs import

**Comment by jessechahal** [comment]
> * "Untitled event" issue: I don't see any possible bug there. The only reason for that is that there is nothing in the "summary" field for those events in the JSON data received. As I can't register to GSuite for free (they say it's for free but a credit card number is mandatory...) I can't try to share such calendar with my test google user. If you can share a GSuite calendar with me ([nyugglee@gmail.com](mailto:nyugglee@gmail.com)) I could inspect the difference in the API response when importing.

I have created a new calendar and shared it with you using a GSuite account

**Comment by eneiluj** [comment]
Thanks a lot for the detailed feedback!

Photo count is wrong because making a precise one triggers too many requests as we discussed before. I'll try to adapt the display so it makes more sense :grin:

You're right about the `fclose()` problem. It's done by the http client. I'll remove the extra useless ones.

> When i try to open the files they fail to open complaining about issues with URL (local ipv4 instead of DNS record)

Not sure I get what you mean. Are you talking about opening documents in Collabora/OnlyOffice? What if you download those problematic files (once they've been imported in Nextcloud). Are they ok? Anyway it seems to be a problem on your side.

> think i found the issue, Gsuite publicly shared calendars by default now only show if busy/free

Ok great. I'll check the content of the API response when importing such calendars and try to detect this precise case to fill the empty titles with something more meaningful.

To be continued

**Comment by eneiluj** [comment]
v0.0.20 is out there. It fixes the fclose problem, set a fallback title for private events (that's what the API says about those I get from your shared calendar), and does not show photo import progress percentage because we don't know the exact number of photos anyway.

**Comment by jessechahal** [comment]
tested with `v0.0.21`
- calendar import of privately shared GSuite calendars looks good
- photos import: no more fclose() errors in logs
- drive import:
  - i probably have something affecting my install of collabera CODE thats causing imported or uploaded files to use server IPv4 address instead of configured Domain name. Probably a bug on their part but maybe something i can fix/workaround on my end somehow. I've looked around online for people who have had similar problems with Nextcloud but every forum post I found never had a solution only unanswered question or recommending upgrade/reinstall nextcloud.
